### PR TITLE
Maunope/network added firewall metrics

### PR DESCRIPTION
**Pull request by maunope**

### CODE REVIEW

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/vpc_firewalls.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+          f"Could not write VPC firewal rules to metric for projects/{project} due to missing quotas"
+      )
+      continue
+
+    network_dict = networks.get_networks(config, project)
+
+    project_usage = 0
+    for net in network_dict:
+      usage = 0
+      if net['network_name'] in firewalls_dict:
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/cloud-function/metrics/vpc_firewalls.py:91]: As discussed, I think there is an issue around this point to be troubleshooted.

**Review by aurelienlegrand — CHANGES_REQUESTED** [review]
Thanks Maurizio for the contribution!

I added a few comments, mostly naming and documentation related.
Also I tested the code and I am waiting for a fix before I merge this PR :)

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/vpc_firewalls.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        config, project, current_quota_limit['limit'],
+        metrics_dict["metrics_per_project"][f"firewalls"]["limit"]["name"])
+    metrics.write_data_to_metric(
+        config, project, project_usage / current_quota_limit['limit']
+        if current_quota_limit['limit'] != 0 else 0,
+        metrics_dict["metrics_per_project"][f"firewalls"]["utilization"]
+        ["name"])
+
+    print(
+        f"Wrote number of VPC Firewall Rules  to metric for projects/{project}")
```
[the remark this answers is not in the document]
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/vpc_firewalls.py:110]: k

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics.yaml [hunk 3]
```diff
@@ -159,4 +159,16 @@ metrics_per_peering_group:
...
-      description: Number of Dynamic routes per peering group - utilization.
\ No newline at end of file
+      description: Number of Dynamic routes per peering group - utilization.
+metrics_per_project:
+  firewalls:
+    usage:
+      name: firewalls_per_project_vpc_usage
+      description: Number of VPC firewall rules in a project - usage.
+    limit:
+      #no default limit as we can assume quotas can be always read from gcloud API
```
[the remark this answers is not in the document]
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics.yaml:169]: k

File: blueprints/cloud-operations/network-dashboard/cloud-function/main.py [hunk 4]
```diff
@@ -33,7 +33,7 @@ def get_monitored_projects_list(config):
         monitored_projects (List of strings): Full list of projects to be monitored
     '''
   monitored_projects = config["monitored_projects"]
-  monitored_folders = os.environ.get("MONITORED_FOLDERS_LIST").split(",")
+  monitored_folders = []  #os.environ.get("MONITORED_FOLDERS_LIST").split(",")
```
[the remark this answers is not in the document]
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/main.py:36]: y

File: blueprints/cloud-operations/network-dashboard/cloud-function/metrics/vpc_firewalls.py (hunk 1 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/vpc_firewalls.py]: fixed, you were right, it's firewalls_dict['project']

Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/cloud-function/metrics/vpc_firewalls.py]: Fixed, you were right, it's firewalls_dict['project']

File: blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json [hunk 5]
```diff
@@ -375,31 +411,34 @@
...
         },
         "width": 6,
-        "yPos": 20
+        "xPos": 6,
+        "yPos": 16
       },
       {
         "height": 4,
         "widget": {
-          "title": "ip_addresses_per_subnet_utilization",
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json:383]: Could you please add the ip_addresses_per_subnet_utilization metric in the dashboard that you removed?

File: blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json [hunk 6]
```diff
@@ -375,31 +411,34 @@
...
+          "title": "firewalls_per_project_usage",
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json:420]: This should be firewalls_per_project_vpc_usage

File: blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json [hunk 7]
```diff
@@ -375,31 +411,34 @@
...
+                      "groupByFields": [
+                        "metric.label.\"project\""
+                      ],
                       "perSeriesAligner": "ALIGN_MEAN"
                     },
-                    "filter": "metric.type=\"custom.googleapis.com/ip_addresses_per_subnet_utilization\" resource.type=\"global\"",
-                    "secondaryAggregation": {
-                      "alignmentPeriod": "60s"
-                    }
+                    "filter": "metric.type=\"custom.googleapis.com/firewalls_per_project_usage\" resource.type=\"global\""
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json:441]: I think the metric should be "firewalls_per_project_vpc_usage":
                    "filter": "metric.type=\"custom.googleapis.com/firewalls_per_project_vpc_usage\" resource.type=\"global\""

File: blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json (hunk 6 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json]: y

File: blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json (hunk 7 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json]: y

File: blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json [hunk 8]
```diff
@@ -429,15 +505,14 @@
                 "plotType": "LINE",
                 "targetAxis": "Y1",
                 "timeSeriesQuery": {
+                  "apiSource": "DEFAULT_CLOUD",
                   "timeSeriesFilter": {
                     "aggregation": {
                       "alignmentPeriod": "60s",
+                      "crossSeriesReducer": "REDUCE_NONE",
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json:512]: Could you please remove that line?

File: blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json [hunk 9]
```diff
@@ -375,12 +411,53 @@
...
+        "widget": {
+          "title": "firewalls_per_project_vpc_usage",
+          "xyChart": {
+            "chartOptions": {
+              "mode": "COLOR"
+            },
+            "dataSets": [
+              {
+                "minAlignmentPeriod": "60s",
+                "plotType": "STACKED_BAR",
```
Comment (aurelienlegrand) [inline blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json:428]: Can you use the following instead (like the other charts):
"plotType": "LINE",

Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json:428]: y

File: blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json (hunk 8 above)
Comment (maunope) [reply blueprints/cloud-operations/network-dashboard/dashboards/quotas-utilization.json:512]: y

**Review by aurelienlegrand — APPROVED** [review]

**Comment by aurelienlegrand** [comment]
I think the PR is good now, thanks a lot for your contribution Maurizio!
LGTM

**Pull request merged**